Safe Transmute: Enable handling references

bryangarza opened the pull request:
This patch enables support for references in Safe Transmute, by generating nested obligations during trait selection. Specifically, when we call `confirm_transmutability_candidate(...)`, we now recursively traverse the `rustc_transmute::Answer` tree and create obligations for all the `Answer` variants, some of which include multiple nested `Answer`s.

bryangarza committed:
Safe Transmute: Enable handling references, including recursive types

This patch enables support for references in Safe Transmute, by generating
nested obligations during trait selection. Specifically, when we call
`confirm_transmutability_candidate(...)`, we now recursively traverse the
`rustc_transmute::Answer` tree and create obligations for all the `Answer`
variants, some of which include multiple nested `Answer`s.

Also, to handle recursive types, enable support for coinduction for the Safe
Transmute trait (`BikeshedIntrinsicFrom`) by adding the `#[rustc_coinduction]`
annotation.

Also fix some small logic issues when reducing the `or` and `and` combinations
in `rustc_transmute`, so that we don't end up with additional redundant
`Answer`s in the tree.

Co-authored-by: Jack Wrenn <jack@wrenn.fyi>

bryangarza commented:
r? @compiler-errors

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4769727132/jobs/8480333477) [(plain)](https://github.com/rust-lang/rust/commit/db58ec11308ae34b00a1cc80ec994ea57643696f/checks/12938274965/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] tests/ui/transmutability/region-infer.rs stdout ----
diff of stderr:

- error[E0277]: `()` cannot be safely transmuted into `W<'_>` in the defining scope of `Context`
-   --> $DIR/region-infer.rs:20:5
- LL |     test();
- LL |     test();
-    |     ^^^^ `W<'_>` does not have a well-specified layout
- note: required by a bound in `test`
- note: required by a bound in `test`
+ error[E0283]: type annotations needed: cannot satisfy `W<'a>: BikeshedIntrinsicFrom<(), Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>`
9    |
9    |
- LL |   fn test<'a>()
- LL |   where
- LL |   where
13 LL |       W<'a>: BikeshedIntrinsicFrom<
15 LL | |             (),

16 LL | |             Context,
16 LL | |             Context,
17 LL | |             { Assume { alignment: true, lifetimes: true, safety: true, validity: true } },
18 LL | |         >,
-    | |_________^ required by this bound in `test`
+    |
+    |
+    = note: cannot satisfy `W<'a>: BikeshedIntrinsicFrom<(), Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>`
21 error: aborting due to previous error
22 

- For more information about this error, try `rustc --explain E0277`.
---
To only update this specific test, also pass `--test-args transmutability/region-infer.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/transmutability/region-infer.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/transmutability/region-infer" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/transmutability/region-infer/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0283]: type annotations needed: cannot satisfy `W<'a>: BikeshedIntrinsicFrom<(), Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>`
  --> fake-test-src-base/transmutability/region-infer.rs:11:12
   |
LL |       W<'a>: BikeshedIntrinsicFrom<
LL | |             (),
LL | |             Context,
LL | |             Context,
LL | |             { Assume { alignment: true, lifetimes: true, safety: true, validity: true } },
LL | |         >,
   |
   |
   = note: cannot satisfy `W<'a>: BikeshedIntrinsicFrom<(), Context, Assume { alignment: true, lifetimes: true, safety: true, validity: true }>`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0283`.
------------------------------------------
```

</details>

bryangarza commented:
Ah, I should have run the rest of the tests
@rustbot author

bryangarza commented:
Regarding the enablement of coinduction for Safe Transmute, I discussed it a bit here https://rust-lang.zulipchat.com/#narrow/stream/144729-t-types/topic/Coinductive.20solving.20for.20.28safe.20transmute.29.20and.20all.20traits/near/351739918

compiler-errors commented on tests/ui/transmutability/arrays/should_require_well_defined_layout.stderr:
```diff
@@ -23,7 +23,7 @@ error[E0277]: `u128` cannot be safely transmuted into `[String; 0]` in the defin
    |
 LL |         assert::is_maybe_transmutable::<u128, repr_rust>();
-   |                                               ^^^^^^^^^ `[String; 0]` does not have a well-specified layout
+   |                                               ^^^^^^^^^ `u128` does not have a well-specified layout
```
what's going on here?

bryangarza replied:
I think the fact that we now reduce the `Answer` tree in a slightly different order, made it so that some errors are bubbling up instead of other ones, but I need to look into it more

bryangarza commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -302,9 +303,55 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+        Ok(ImplSourceBuiltinData { nested })
+    }
+
+    fn smoosh(
```
haha will do

bryangarza commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -302,9 +303,55 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+            rustc_transmute::Answer::No(_) => Err(Unimplemented),
+            rustc_transmute::Answer::IfAll(obs) | rustc_transmute::Answer::IfAny(obs) => {
+                let mut nested = vec![];
+                for smooshed in obs.into_iter().map(smoosh_curried) {
+                    nested.extend(smooshed?);
```
yep, it was just a placeholder while we were iterating on the code, will rename

bryangarza commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -302,9 +303,55 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                };
+
+                // // FIXME(bryangarza): Check src.mutability or dst.mutability to know whether dst -> src obligation is needed
+                Ok(vec![make_obl(src.ty, dst.ty), make_obl(dst.ty, src.ty)])
```
According to @jswrenn, it should be checked bidirectionally if Dst is mutable, for example, if you go from bool to u8, it's okay as long as you can't update that u8 to be > 1. Something like that.

compiler-errors replied:
Makes sense.

bryangarza commented on compiler/rustc_transmute/src/layout/mod.rs:
```diff
@@ -30,33 +30,47 @@ impl fmt::Debug for Byte {
-    impl<'tcx> super::Ref for Ref<'tcx> {}
+    impl<'tcx> super::Ref for Ref<'tcx> {
+        fn min_align(&self) -> usize {
+            // FIXME(bryangarza): Need to have access to a ctx: TyCtxt<'tcx> here
+            // super::tree::rustc::layout_of(ctx, self.ty).unwrap().align()
```
Yeah, the main problem I have here is that the code is split by a feature flag that enables rustc dependencies, and here I only have access to a trait object `QueryContext` that is supposed to abstract `TyCtxt` so it's not directly used... IDK what the best way to deal with this is, downcasting isn't really possible because we don't have a static lifetime for `TyCtxt`, the other way I could handle it is just moving more code to use the rustc feature flag but at that point, we should maybe just get rid of the flag and always depend on rustc. (IDK why, but I think the original reason we have it is to be able to build `rustc_transmute` as its own crate independent of the compiler.

compiler-errors replied:
Well then QueryContext would be the thing that has a layout-of method exposed on it, it seems?

But yeah, I don't think all this extra abstraction is actually helping the design of this crate. I'd personally rather this be designed with rustc in mind, then later if we want this to be a separate crate (like Chalk) then it can be re-abstracted out with a clearer understanding of what the API boundaries need to be.

jswrenn replied:
How about we just add an `align` *field* to `Ref`, and fill in that field during the`Tree::from_ty` process (where we have access to `TyCtxt`).

bryangarza commented on library/core/src/mem/transmutability.rs:
```diff
@@ -5,6 +5,7 @@
 /// notwithstanding whatever safety checks you have asked the compiler to [`Assume`] are satisfied.
 #[unstable(feature = "transmutability", issue = "99571")]
 #[lang = "transmute_trait"]
+#[rustc_coinductive]
```
Yeah, will open one soon

compiler-errors replied:
FCPs are opened by team members. I can open one when this PR is ready to land.

jswrenn commented on compiler/rustc_transmute/src/maybe_transmutable/mod.rs:
```diff
@@ -214,34 +214,99 @@ where
+                                    .into_iter()
+                                    .map(|(&dst_ref, &dst_state_prime)| {
+                                        if !src_ref.is_mutable() && dst_ref.is_mutable() {
+                                            Answer::No(Reason::MutabilityRelaxation)
```
```suggestion
                                            Answer::No(Reason::DstIsMoreUnique)
```

jswrenn commented on compiler/rustc_transmute/src/lib.rs:
```diff
@@ -54,6 +51,10 @@ pub enum Reason {
     /// `Dst` is larger than `Src`, and the excess bytes were not exclusively uninitialized.
     DstIsTooBig,
+    /// Src has to have a stricter algignment than Dst, but it does not.
+    InvalidAlignment,
```
Keeping with the pattern of variant names that are a little more explanatory:
```suggestion
    DstHasStricterAlignment,
```

jswrenn commented on compiler/rustc_transmute/src/maybe_transmutable/mod.rs:
```diff
@@ -214,34 +214,99 @@ where
+                                        } else if !self.assume.alignment
+                                            && src_ref.min_align() < dst_ref.min_align()
+                                        {
+                                            Answer::No(Reason::InvalidAlignment)
```
```suggestion
                                            Answer::No(Reason::DstHasStricterAlignment)
```

jswrenn commented on compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:
```diff
@@ -2967,6 +2967,14 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     rustc_transmute::Reason::DstIsTooBig => {
                         format!("The size of `{src}` is smaller than the size of `{dst}`")
                     }
+                    rustc_transmute::Reason::InvalidAlignment => {
```
```suggestion
                    rustc_transmute::Reason::DstHasStricterAlignment => {
```

bryangarza committed:
Safe Transmute: Change Answer type to Result

This patch updates the `Answer` type from `rustc_transmute` so that it just a
type alias to `Result`. This makes it so that the standard methods for `Result`
can be used to process the `Answer` tree, including being able to make use of
the `?` operator on `Answer`s.

Also, remove some unused functions

bryangarza commented on tests/ui/transmutability/region-infer.stderr:
```diff
@@ -2,7 +2,7 @@ error[E0277]: `()` cannot be safely transmuted into `W<'_>` in the defining scop
   --> $DIR/region-infer.rs:20:5
    |
 LL |     test();
-   |     ^^^^ `W<'_>` does not have a well-specified layout
```
I *think* this change is expected because the previous error was because we weren't supporting references.

This is what is in the test
```rs
#[repr(C)]
struct W<'a>(&'a ());
```

bryangarza replied:
@jswrenn do you agree with that^?

bryangarza replied:
Thinking about it more, the new error is
```
The size of `()` is smaller than the size of `W<'_>`
```
which makes sense because `W` needs to store the ptr. So `()` does not hold enough data to be able to transmute to `W`. Just going to resolve this comment for now!

bryangarza committed:
Safe Transmute: Fix propagation of errors

- Make sure that the most specific Reason is the one that bubbles up when we
  are folding over the `Answer` tree. `Reason::DstIsBitIncompatible` is the
  least specific, so that should be used only when there isn't anything else
  available.
- Small fixes where we used the wrong Reason variant.
- Tiny cleanups

bryangarza commented on tests/ui/transmutability/enums/repr/primitive_reprs_should_have_correct_length.stderr:
```diff
@@ -90,7 +90,7 @@ error[E0277]: `u8` cannot be safely transmuted into `V0i16` in the defining scop
   --> $DIR/primitive_reprs_should_have_correct_length.rs:72:44
    |
 LL |         assert::is_transmutable::<Smaller, Current, Context>();
-   |                                            ^^^^^^^ At least one value of `u8` isn't a bit-valid value of `V0i16`
```
The changes in this file are to be expected, I made it so that `DstIsBitIncompatible` is only used if there are no other errors that are more specific

bryangarza committed:
Safe Transmute: Add alignment tests

bryangarza committed:
Safe Transmute: Check mutability before creating dst -> src obligation

- Only create dst -> src obligation if Dst is mutable
- Add some long comments to explain parts of the transmutability code that were
  unclear to me when reading
- Update/add tests

bryangarza commented:
@rustbot ready

jswrenn commented on compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:
```diff
@@ -2935,13 +2935,15 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
             rustc_transmute::Assume::from_const(self.infcx.tcx, obligation.param_env, trait_ref.substs.const_at(3)) else {
                 span_bug!(span, "Unable to construct rustc_transmute::Assume where it was previously possible");
             };
+        // FIXME(bryangarza): Is this enough, or should we resolve all nested
+        // obligations like we do for `confirm_transmutability_candidate(...)?`
```
There shouldn't be any circumstances under which this is producing nested obligations. If we're here, it's because the result we got back was an `Err`.

bryangarza replied:
But couldn't the original call to confirm_transmutability_candidate have had the Err somewhere in the nested obligations? Remember, we are checking transmutability from scratch again here, so if we get an Answer tree, it's not flattened yet

bryangarza replied:
I'll remove this for now and if we seem to run into issues we can just address it, otherwise the comment will just be confusing for future readers...

jswrenn commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -269,11 +270,67 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+            match answer {
+                Ok(None) => Ok(vec![]),
+                Err(_) => Err(Unimplemented),
+                // FIXME(bryangarza): Add separate `IfAny` case, instead of treating as `IfAll`
```
An important note that should be added here: This is probably blocked on the trait solver needing to support disjunctions of obligations.

bryangarza replied:
Yeah, that's something I was planning on tackling at some point -- will leave a FIXME

jswrenn commented on compiler/rustc_transmute/src/layout/mod.rs:
```diff
@@ -30,33 +30,46 @@ impl fmt::Debug for Byte {
 pub(crate) trait Def: Debug + Hash + Eq + PartialEq + Copy + Clone {}
-pub trait Ref: Debug + Hash + Eq + PartialEq + Copy + Clone {}
+pub trait Ref: Debug + Hash + Eq + PartialEq + Copy + Clone {
+    fn min_align(&self) -> usize {
+        1
+    }
+
+    fn is_mutable(&self) -> bool {
+        false
+    }
```
Since this is safety-critical code, let's not have default impls here.

bryangarza replied:
What should we have instead, `unimplemented!`?

jswrenn commented on compiler/rustc_transmute/src/layout/mod.rs:
```diff
@@ -30,33 +30,46 @@ impl fmt::Debug for Byte {
+}
 
 impl Def for ! {}
 impl Ref for ! {}
```
```suggestion
impl Ref for ! {
    fn min_align(&self) -> usize {
        unreachable!()
    }

    fn is_mutable(&self) -> bool {
        unreachable!()
    }
}
```

bryangarza replied:
ah, thanks, yeah I wasn't sure what was the right thing to put here

jswrenn commented on compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:
```diff
@@ -2963,19 +2965,26 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     }
+                    rustc_transmute::Reason::DstHasStricterAlignment => {
+                        format!(
+                            "The alignment of `{src}` should be stricter than that of `{dst}`, but it is not"
```
```suggestion
                            "The minimum alignment of `{src}` should be greater than that of `{dst}`, but it is not"
```

It'd be cool to have `note:`s that said exactly what the minimum alignments of src and dst are, but that can wait for a future PR.

jswrenn commented on compiler/rustc_transmute/src/lib.rs:
```diff
@@ -1,4 +1,4 @@
-#![feature(alloc_layout_extra, decl_macro, iterator_try_reduce, never_type)]
+#![feature(alloc_layout_extra, decl_macro, iterator_try_reduce, never_type, let_chains)]
```
Aren't `let_chains` stable?

bryangarza commented on compiler/rustc_transmute/src/lib.rs:
```diff
@@ -19,18 +19,12 @@ pub struct Assume {
 
+/// A condition which must hold for safe transmutation to be possible
+#[derive(Debug, Hash, Eq, PartialEq, Clone)]
+pub enum Condition<R> {
```
I think that during the eval'ing of the `Answer` tree, it's not until we process the next layer up that we then filter out or bubble up the errors. But let me see if I can move some things around to only hold `Condition`, might be cleaner.

compiler-errors commented:
@rustbot author

bryangarza commented on tests/ui/transmutability/alignment/align-fail.rs:
```diff
@@ -0,0 +1,23 @@
+// check-fail
```
I think it's supposed to fail, the alignment of the source is smaller. I'm going to update the error reporting to actually include the alignments so that we can see what's what

bryangarza committed:
Safe Transmute: Update definition of Condition type

- Change `Condition` to not contain `Answer`s but instead just contain other
  `Condition`s directly.
- Also improve error reporting for `DstHasStricterAlignment`

bryangarza commented:
@rustbot ready

compiler-errors commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -279,11 +280,67 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                    // If Dst is mutable, check bidirectionally.
+                    // For example, transmuting bool -> u8 is OK as long as you can't update that u8
+                    // to be > 1, because you could later transmute the u8 back to a bool and get UB.
+                    let mut obligations = vec![make_obl(src.ty, dst.ty)];
+                    if dst.mutability == Mutability::Mut {
+                        obligations.push(make_obl(dst.ty, src.ty));
+                    }
```
exhaustively match on this, and either produce a `vec[obl(src, dst)]`, or a `vec[obl(..), obl(dst, src)]`

compiler-errors commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -279,11 +280,67 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                rustc_transmute::Condition::IfAll(answers)
+                | rustc_transmute::Condition::IfAny(answers) => {
+                    let mut nested = vec![];
+                    for flattened in answers
+                        .into_iter()
+                        .map(|answer| flatten_answer_tree(tcx, obligation, predicate, answer))
```
into_iter, flat_map, then collect?

compiler-errors commented on compiler/rustc_transmute/src/maybe_transmutable/mod.rs:
```diff
@@ -75,14 +76,14 @@ mod rustc {
-                    (_, Err(Err::Unspecified)) => Err(Answer::No(Reason::DstIsUnspecified)),
+                    (Err(Err::TypeError(_)), _)
+                    | (_, Err(Err::TypeError(_)))
+                    | (Err(Err::Unknown), _)
+                    | (_, Err(Err::Unknown)) => Err(Ok(None)),
```
I don't think it's correct to treat `Unknown` like this. We will eventually reach this when we support transmuting `&T` that references a generic parameter `T`, since if `T` is not sized then it'll have an unknown layout, and I don't think we want to treat that as transmutability?

Side-note, using `Result` as if it were `Either` here is confusing. Maybe it should be lifted into its own variant. Same with the `Result` inside of that `Result`, which is (I think) `Answer<T>` -- why not just keep that as its own enum with different variants corresponding to "`Yes`", "`OnlyIf`", and "`No`"?

Side-note of a side-note: `map_layouts` seems like a really obtuse way of unpacking a `MaybeTransmutableQuery`. We really shouldn't be doing all this, tbh.

compiler-errors replied:
So to summarize, maybe let's get rid of `map_layouts` altogether?

bryangarza replied:
> I don't think it's correct to treat `Unknown` like this. We will eventually reach this when we support transmuting `&T` that references a generic parameter `T`, since if `T` is not sized then it'll have an unknown layout, and I don't think we want to treat that as transmutability?

The purpose of returning `Ok(None)` (previously `Answer::Yes`) is to just exit early and not emit any errors. This is because an error will already be emitted by rustc because of the unknown layout. It's not that we are saying "yes, transmutation is possible", but rather we just quit out here and let rustc emit some other error.

> 
> Side-note, using `Result` as if it were `Either` here is confusing. Maybe it should be lifted into its own variant. Same with the `Result` inside of that `Result`, which is (I think) `Answer<T>` -- why not just keep that as its own enum with different variants corresponding to "`Yes`", "`OnlyIf`", and "`No`"?

Yeah, I agree the nested results is confusing to read. I'll  see if I can improve it.

> 
> Side-note of a side-note: `map_layouts` seems like a really obtuse way of unpacking a `MaybeTransmutableQuery`. We really shouldn't be doing all this, tbh.

For sure, it's not easy to read/understand. Let's refactor it in a separate PR, I think it will add a bunch more changes to do it in here.

compiler-errors replied:
> The purpose of returning Ok(None) (previously Answer::Yes) is to just exit early and not emit any errors. This is because an error will already be emitted by rustc because of the unknown layout.

Where is that error emitted? Rust will not, in general, emit an error on an *unknown* layout. Maybe I'm misunderstanding something.

compiler-errors replied:
For example, this code relies on layout computation returning "unknown" not emitting an error:

```
#![feature(pointer_like_trait)]

use std::ops::Deref;
use std::marker::PointerLike;

impl<T: PointerLike> Wrapper<T> {
    fn do_it(self) {
        println!("is pointer-like");
    }
}

impl WrapperDeref {
    fn do_it(&self) {
        println!("is not pointer-like");
    }
}

struct Wrapper<T>(T);

impl<T> Deref for Wrapper<T> {
    type Target = WrapperDeref;
    
    fn deref(&self) -> &WrapperDeref { &WrapperDeref }
}

struct WrapperDeref;

fn test<T>(t: T) {
    Wrapper(&t).do_it();
    Wrapper(t).do_it();
}

fn main() {
    test(1i32);
}
```

Inside of the method `test`, we first check if `T: PointerLike` holds, which is going to return `LayoutError::Unknown`, which is *not* a fatal error.

bryangarza replied:
Hmm okay thanks for the detailed comment, I guess I should change the `Unknown` case to emit an error then...

@jswrenn do you remember why you had left the comment on line 79 saying that `Unknown` would already be reported?

compiler-errors replied:
It's not *currently* reachable because we don't even check transmutability for things that need substitution (which I assume could get lifted). 

But I assume it will, and I'm not confident that we'll remember to set this to fail the predicate, and it'll just cause the transmutability predicate to succeed.

compiler-errors commented on compiler/rustc_transmute/src/maybe_transmutable/mod.rs:
```diff
@@ -203,119 +203,237 @@ where
-                Self::IfAll(constraints)
-            }
-            (lhs, rhs) => Self::IfAll(vec![lhs, rhs]),
+    // If both are errors, then we should return the more specific one
```
I think all of this logic could be simplified into something like:

```
match (lhs, rhs) {
  (Err(Reason::DstIsBitIncompatible), _) => lhs,
  (Err(_), _) => lhs,
  ...,
  (Ok(None), Ok(other)) => Ok(other),
  ...,
}
```

compiler-errors commented on compiler/rustc_transmute/src/maybe_transmutable/mod.rs:
```diff
@@ -203,119 +203,237 @@ where
-        });
-    result
+    // If both are errors, then we should return the more specific one
+    if lhs.is_err() && rhs.is_err() {
```
Similarly

bryangarza committed:
Safe Transmute: Disable coinduction support

This patch just removes the `#[rustc_coinductive]` annotation from `BikeshedIntrinsicFrom` and flips the related tests to `check-fail`. Once an FCP for setting the annotation on the trait is approved, it could be enabled again.

bryangarza committed:
Safe Transmute: Refactor error handling and Answer type

- Create `Answer` type that is not just a type alias of `Result`
- Remove a usage of `map_layouts` to make the code easier to read
- Don't hide errors related to Unknown Layout when computing transmutability

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5249608488/jobs/9482591861) [(plain)](https://github.com/rust-lang/rust/commit/ef0b78c7da05d4a9f302e8f1905ef15c7b6c43ae/checks/14205080908/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
36 | macro_rules! assert_eq {
   | ---------------------- in this expansion of `assert_eq!`
...
40 |                 if !(*left_val == *right_val) {
   |                      ---------    ^^^^^^^^^^ expected `Answer<!>`, found `Result<Option<_>, _>`
   |                      |
   |                      expected because this is `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:20:9
   |
   |
20 |         assert_eq!(answer, Ok(None));
   |
   |
   = note: expected enum `Answer<!>`
              found enum `Result<std::option::Option<_>, _>`
error[E0308]: mismatched types
  --> /checkout/library/core/src/macros/mod.rs:40:35
   |
36 | macro_rules! assert_eq {
36 | macro_rules! assert_eq {
   | ---------------------- in this expansion of `assert_eq!`
...
40 |                 if !(*left_val == *right_val) {
   |                      ---------    ^^^^^^^^^^ expected `Answer<!>`, found `Result<Option<_>, _>`
   |                      |
   |                      expected because this is `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:33:9
   |
   |
33 |         assert_eq!(answer, Ok(None));
   |
   |
   = note: expected enum `Answer<!>`
              found enum `Result<std::option::Option<_>, _>`
error[E0308]: mismatched types
  --> /checkout/library/core/src/macros/mod.rs:53:35
   |
36 |   macro_rules! assert_eq {
36 |   macro_rules! assert_eq {
   |   ---------------------- in this expansion of `assert_eq!`
...
53 |                   if !(*left_val == *right_val) {
   |                                     ^^^^^^^^^^ expected `Result<Option<_>, _>`, found `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:67:21
   |
67 | /                     assert_eq!(
68 | |                         Ok(None),
68 | |                         Ok(None),
69 | |                         MaybeTransmutableQuery::new(
70 | |                             src_layout.clone(),
...  |
79 | |                         dst_layout
80 | |                     );
   | |_____________________- in this macro invocation
   |
   = note: expected enum `Result<std::option::Option<_>, _>`
              found enum `Answer<!>`
   |
   |
53 |                 if !(*left_val == Err(*right_val)) {

error[E0308]: mismatched types
  --> /checkout/library/core/src/macros/mod.rs:53:35
   |
   |
36 |   macro_rules! assert_eq {
   |   ---------------------- in this expansion of `assert_eq!`
...
53 |                   if !(*left_val == *right_val) {
   |                                     ^^^^^^^^^^ expected `Result<Option<_>, _>`, found `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:82:21
   |
82 | /                     assert_eq!(
83 | |                         Ok(None),
83 | |                         Ok(None),
84 | |                         MaybeTransmutableQuery::new(
85 | |                             src_layout.clone(),
...  |
94 | |                         dst_layout
95 | |                     );
   | |_____________________- in this macro invocation
   |
   = note: expected enum `Result<std::option::Option<_>, _>`
              found enum `Answer<!>`
   |
   |
53 |                 if !(*left_val == Err(*right_val)) {

error[E0308]: mismatched types
   --> /checkout/library/core/src/macros/mod.rs:53:35
    |
    |
36  |   macro_rules! assert_eq {
    |   ---------------------- in this expansion of `assert_eq!`
...
53  |                   if !(*left_val == *right_val) {
    |                                     ^^^^^^^^^^ expected `Result<_, Reason>`, found `Answer<!>`
   ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:97:21
    |
97  | /                     assert_eq!(
97  | /                     assert_eq!(
98  | |                         Err(Reason::DstIsBitIncompatible),
99  | |                         MaybeTransmutableQuery::new(
100 | |                             src_layout.clone(),
109 | |                         dst_layout
110 | |                     );
    | |_____________________- in this macro invocation
    |
    |
    = note: expected enum `Result<_, Reason>`
               found enum `Answer<!>`
help: try wrapping the expression in `Ok`
    |
53  |                 if !(*left_val == Ok(*right_val)) {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_transmute` (lib test) due to 5 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5249646794/jobs/9482676470) [(plain)](https://github.com/rust-lang/rust/commit/1556d779b50074bf8a8e1a346cd0669d33262365/checks/14205189786/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
36 | macro_rules! assert_eq {
   | ---------------------- in this expansion of `assert_eq!`
...
40 |                 if !(*left_val == *right_val) {
   |                      ---------    ^^^^^^^^^^ expected `Answer<!>`, found `Result<Option<_>, _>`
   |                      |
   |                      expected because this is `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:20:9
   |
   |
20 |         assert_eq!(answer, Ok(None));
   |
   |
   = note: expected enum `Answer<!>`
              found enum `Result<std::option::Option<_>, _>`
error[E0308]: mismatched types
  --> /checkout/library/core/src/macros/mod.rs:40:35
   |
36 | macro_rules! assert_eq {
36 | macro_rules! assert_eq {
   | ---------------------- in this expansion of `assert_eq!`
...
40 |                 if !(*left_val == *right_val) {
   |                      ---------    ^^^^^^^^^^ expected `Answer<!>`, found `Result<Option<_>, _>`
   |                      |
   |                      expected because this is `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:33:9
   |
   |
33 |         assert_eq!(answer, Ok(None));
   |
   |
   = note: expected enum `Answer<!>`
              found enum `Result<std::option::Option<_>, _>`
error[E0308]: mismatched types
  --> /checkout/library/core/src/macros/mod.rs:53:35
   |
36 |   macro_rules! assert_eq {
36 |   macro_rules! assert_eq {
   |   ---------------------- in this expansion of `assert_eq!`
...
53 |                   if !(*left_val == *right_val) {
   |                                     ^^^^^^^^^^ expected `Result<Option<_>, _>`, found `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:67:21
   |
67 | /                     assert_eq!(
68 | |                         Ok(None),
68 | |                         Ok(None),
69 | |                         MaybeTransmutableQuery::new(
70 | |                             src_layout.clone(),
...  |
79 | |                         dst_layout
80 | |                     );
   | |_____________________- in this macro invocation
   |
   = note: expected enum `Result<std::option::Option<_>, _>`
              found enum `Answer<!>`
   |
   |
53 |                 if !(*left_val == Err(*right_val)) {

error[E0308]: mismatched types
  --> /checkout/library/core/src/macros/mod.rs:53:35
   |
   |
36 |   macro_rules! assert_eq {
   |   ---------------------- in this expansion of `assert_eq!`
...
53 |                   if !(*left_val == *right_val) {
   |                                     ^^^^^^^^^^ expected `Result<Option<_>, _>`, found `Answer<!>`
  ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:82:21
   |
82 | /                     assert_eq!(
83 | |                         Ok(None),
83 | |                         Ok(None),
84 | |                         MaybeTransmutableQuery::new(
85 | |                             src_layout.clone(),
...  |
94 | |                         dst_layout
95 | |                     );
   | |_____________________- in this macro invocation
   |
   = note: expected enum `Result<std::option::Option<_>, _>`
              found enum `Answer<!>`
   |
   |
53 |                 if !(*left_val == Err(*right_val)) {

error[E0308]: mismatched types
   --> /checkout/library/core/src/macros/mod.rs:53:35
    |
    |
36  |   macro_rules! assert_eq {
    |   ---------------------- in this expansion of `assert_eq!`
...
53  |                   if !(*left_val == *right_val) {
    |                                     ^^^^^^^^^^ expected `Result<_, Reason>`, found `Answer<!>`
   ::: compiler/rustc_transmute/src/maybe_transmutable/tests.rs:97:21
    |
97  | /                     assert_eq!(
97  | /                     assert_eq!(
98  | |                         Err(Reason::DstIsBitIncompatible),
99  | |                         MaybeTransmutableQuery::new(
100 | |                             src_layout.clone(),
109 | |                         dst_layout
110 | |                     );
    | |_____________________- in this macro invocation
    |
    |
    = note: expected enum `Result<_, Reason>`
               found enum `Answer<!>`
help: try wrapping the expression in `Ok`
    |
53  |                 if !(*left_val == Ok(*right_val)) {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_transmute` (lib test) due to 5 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

bryangarza commented:
@rustbot ready

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit f4cf8f65a5e8e110c8c36469d31f16e8571e2c1a has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors f4cf8f65a5e8e110c8c36469d31f16e8571e2c1a -->
<!-- homu: {"type":"Approved","sha":"f4cf8f65a5e8e110c8c36469d31f16e8571e2c1a","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5264810949/jobs/9520486619)
Approved by: compiler-errors
Pushing 3ed2a10d173d6c2e0232776af338ca7d080b1cd4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5264810949/jobs/9520486619"},"merge_sha":"3ed2a10d173d6c2e0232776af338ca7d080b1cd4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5264810949/jobs/9520486619)
Approved by: compiler-errors
Pushing 3ed2a10d173d6c2e0232776af338ca7d080b1cd4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5264810949/jobs/9520486619"},"merge_sha":"3ed2a10d173d6c2e0232776af338ca7d080b1cd4"} -->

rust-timer commented:
Finished benchmarking commit (3ed2a10d173d6c2e0232776af338ca7d080b1cd4): [comparison URL](https://perf.rust-lang.org/compare.html?start=57c215b08e18054c64428e00a291b45dc690d9de&end=3ed2a10d173d6c2e0232776af338ca7d080b1cd4&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=57c215b08e18054c64428e00a291b45dc690d9de&end=3ed2a10d173d6c2e0232776af338ca7d080b1cd4&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=57c215b08e18054c64428e00a291b45dc690d9de&end=3ed2a10d173d6c2e0232776af338ca7d080b1cd4&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [2.2%, 2.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -3.3% | [-3.3%, -3.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.5% | [-3.3%, 0.3%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=57c215b08e18054c64428e00a291b45dc690d9de&end=3ed2a10d173d6c2e0232776af338ca7d080b1cd4&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.9% | [-6.0%, -3.9%] | 5     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=57c215b08e18054c64428e00a291b45dc690d9de&end=3ed2a10d173d6c2e0232776af338ca7d080b1cd4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 647.762s -> 648.428s (0.10%)
